enhancements/ingress/aws-load-balancer-operator.md

```diff
@@ -0,0 +1,394 @@
+aws-load-balancer will henceforth be referred to as **“lb-controller”** for the
+sake of brevity.
+
+## Motivation
+
+Implement an operator, tentatively called `aws-load-balancer-operator`, which
+deploys and manages an instance of the lb-operator. The operator will be
+distributed through Operator Hub.
+
+### Goals
+* Implement operator which can be installed and used in OCP/OKD.
+
+### Non-Goals
+* Implementing support for ALB for Openshift Routes
+* Support for multiple installations of the controller i.e. multiple custom
+  resources.
+* Support for `ip` mode for Ingress resources
+* Support for NLB backed Services through the lb-controller.
+
+## Proposal
+
+### User Stories
+
+#### TLS Termination and DDoS protection
+
+The user wants to route traffic for the domain example.com to pods of Service
+example-service and have TLS termination and DDoS protection on the load
+balancer. The user deploys the aws-alb-operator and creates an instance of the
+resource AWSLoadBalancerController as follows:
+
+```yaml
+kind: AWSLoadBalancerController
+group: networking.openshift.io/v1beta1
+metadata:
+  name: cluster
+spec:
+  autoTagSubnet: true
+  ingressClass: tls-termination
+```
+Then they create an ingress resource with the following schema:
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: example
+  annotations:
+    alb.ingress.kubernetes.io/scheme: internet-facing
+    alb.ingress.kubernetes.io/certificate-arn: arn:aws:acm:us-west-2:xxxxx
+    alb.ingress.kubernetes.io/shield-advanced-protection: true
+spec:
+  ingressClassName: tls-termination
+  rules:
+  - host: example.com
+    http:
+        paths:
+          - path: /
+            pathType: Exact
+            backend:
+              service:
+                name: example-service
+                port:
+                  number: 80
+```
+The controller then creates an application load balancer with an HTTPS listener
+which terminates the TLS connection with the certificate ARN which is specified
+in the annotation. If the certificate does not exist, the load balancer is not
+created. It also enables AWS Shield on the Elastic IP associated with the ALB.
+
+#### Multiple Ingress through single ALB
+
+The user wants to route traffic to services example-1, example-2 and example-3
+as parts of the domain example.com and wants all the endpoints served through a
+single ALB. The user create an `AWSLoadBalancerController` like in the previous example and
+then creates 3 Ingress resources as follows:
+
+```yaml
+kind: Ingress
+metadata:
+  name: example-1
+  annotations:
+    alb.ingress.kubernetes.io/scheme: internet-facing
+    alb.ingress.kubernetes.io/group.name: example
+    alb.ingress.kubernetes.io/group.order: “1”
+spec:
+  ingressClass: alb
+  rules:
+  - host: example.com
+    http:
+        paths:
+        - path: /blog
+          backend:
+            service:
+        name: example-1
+        port:
+          number: 80
```

Comment:
Indentation is wrong for this `service:` stanza, and likewise for the two ingresses below.